Feature in app updater

notaspirit opened the pull request:
# Feature in app updater

**Implemented:**
- in app updater to replace previous version check + manual install
- additional setting to still allow checking for updates on startup with the user having to agree to update

dragonzkiller commented on WolvenKit.App/Services/ISettingsManager.cs:
```diff
@@ -80,6 +80,17 @@ public static string GetWorkDir()
         return dir;
     }
 
+    public static string GetTemp_DownloadsPath()
```
You should use `System.IO.Path.GetTempPath()` for doing temp things.
1. That's why it's there
2. So if something happens during the update, they can always clear any junk left behind in the Windows Disk Cleanup

notaspirit replied:
changed that in the latest commit

dragonzkiller commented on WolvenKit.App/Services/IUpdateService.cs:
```diff
@@ -0,0 +1,10 @@
+
+public interface IUpdateService
+{
+    public Task<bool> IsUpdateAvailable();
```
We had similar functions in `AppViewModel` on wkit startup, is this what you're removing from there and reimplementing as a service here?

dragonzkiller commented on WolvenKit.App/Services/UpdateService.cs:
```diff
@@ -0,0 +1,151 @@
+        ZipFile.ExtractToDirectory(downloadZipPath, unzipPath);
+        
+        var exePath = Environment.ProcessPath ?? Path.Combine(AppContext.BaseDirectory, "WolvenKit.exe");
+        var scriptPath = Path.Combine(ISettingsManager.GetTemp_DownloadsPath(), "update.ps1");
```
Why do we need both a Powershell and VBS to perform the update? I understand offloading the download and update to a separate process so you can access the locked files, but why both?

dragonzkiller commented on WolvenKit.App/Services/UpdateService.cs:
```diff
@@ -0,0 +1,151 @@
+        File.WriteAllText(vbsScriptPath, $@"Set objShell = CreateObject(""WScript.Shell"")
+objShell.Run ""powershell.exe -ExecutionPolicy Bypass -File """"{scriptPath}"""""", 0, False
+");
+        Process.Start(new ProcessStartInfo
```
Have you tested this on systems where the Execution Policy for Powershell scripts was at a higher security level?

dragonzkiller replied:
Hmmm I see that now. While I know it seems advantageous for us to use it, this would seem like a pretty big security flaw if that's the case. Seems to be it though... Alright

dragonzkiller commented on WolvenKit.App/Services/UpdateService.cs:
```diff
@@ -0,0 +1,151 @@
+        
+        var unzipPath = Path.Join(ISettingsManager.GetTemp_DownloadsPath(), "unzip");
+        Directory.CreateDirectory(unzipPath);
+        ZipFile.ExtractToDirectory(downloadZipPath, unzipPath);
```
We should be checking the SHA256 digest before we do any unzipping to verify the download.

notaspirit replied:
it now checks that in the latest commit (had to sadly switch away from octokit as it doesn't expose the sha256 which was added somewhat recently)

dragonzkiller reviewed (changes requested):
Admittedly I'm on mobile right now, so my review is limited to a quick look at the moment. I think this currently relies on accepting the scripts will always execute properly. You use C# to download and manage the error handling of the download, but it assumes everything has downloaded and installs properly.

wopss commented on WolvenKit.App/Services/UpdateService.cs:
```diff
@@ -0,0 +1,168 @@
+        var exePath = Environment.ProcessPath ?? Path.Combine(AppContext.BaseDirectory, "WolvenKit.exe");
+        var scriptPath = Path.Combine(tempPath, "update.ps1");
+        var vbsScriptPath = Path.Combine(tempPath, "update.vbs");
+        File.WriteAllText(scriptPath, $@"
```
You should be really careful with this, AV software might flag the whole WKit if you run files like this.

Or, they could even block the execution of these scripts until they are analyzed in their clouds.

notaspirit commented:
It's ready for the next review cycle, changed the scripts out for a separate executable that is part of the release assets and added a simple dialog.

to test this new version change the repo owner in `WolvenKit\WolvenKit.App\Services\UpdateService.cs` Line `180` to `notaspirit` I have used the modified build actions to build release assets like they would be after this is merged (only change is the repo owner so it continues to pull from the test assets)

dragonzkiller commented:
Merged!
